### DISCUSSION TITLE
Identity Broker with OpenID Connect - Subject Identifier delivered in different attribute instead of "sub"

### DISCUSSION THREAD

**Discussion by matrium0** [keycloak/keycloak#19208] [Q&A]
Hey,

we are using Keycloak as an Identity-Broker and it works like a charm, with multiple OpenID-Connect Identity Providers. Sadly one of our IDPs suddenly changed something drastic in their implementation. They decided, that from now on we get the unique-user-identifier ("Subject Identifier" according to OIDC spec) in a custom claim "uq-userid", instead of the usual claim "sub".

To make it more clear, what they are doing looks basically like this in the JWT:
INSTEAD OF
```
{
 "sub": "<subject identifier>"
  ...
}
```
THEY NOW GIVE US
```
{
 "sub": "<RANDOM STRING>",
 "uq-userid": "<subject identifier>"
}
```
This of course breaks with the OpenID Connect specification. I discussed this heavily with them. They do have no good reason to do it, but still will 100% implement it like that.

How can I make Keycloak accept the USER-IDENTIFIER from a different claim? I guess I would have to write some sort of extension, that maps from uq-userid to sub, but even after reading the documentation, I have no clue how to start. Maybe you can point me in into the right direction?

Kind regards,
Mario

**Comment by davux** [comment]
I think you can use mappers for that.

**Reply by danielFesenmeyer** [reply]
@matrium0 It seems you can use a [UsernameTemplateMapper](https://github.com/keycloak/keycloak/blob/0fc4d6120bee90504bea588ad85c3d04774bf9c6/services/src/main/java/org/keycloak/broker/oidc/mappers/UsernameTemplateMapper.java) with the following configuration:

- Template: ${CLAIM.uq-userid}
- Target: BROKER_ID
